extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/graal/DatabaseTypeRegisterSubstitutions.java

```diff
@@ -0,0 +1,70 @@
+package io.quarkus.flyway.runtime.graal;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.flywaydb.core.internal.database.DatabaseType;
+import org.flywaydb.core.internal.database.DatabaseTypeRegister;
+
+import com.oracle.svm.core.annotate.Alias;
+import com.oracle.svm.core.annotate.RecomputeFieldValue;
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+
+@TargetClass(DatabaseTypeRegister.class)
+public final class DatabaseTypeRegisterSubstitutions {
+    @Alias
+    @RecomputeFieldValue(kind = RecomputeFieldValue.Kind.FromAlias)
+    private static List<DatabaseType> registeredDatabaseTypes = new ArrayList<>();
+
+    @Alias
+    @RecomputeFieldValue(kind = RecomputeFieldValue.Kind.FromAlias)
+    private static boolean hasRegisteredDatabaseTypes = false;
+
+    @Substitute
+    private static void registerDatabaseTypes() {
+        synchronized (registeredDatabaseTypes) {
+            if (hasRegisteredDatabaseTypes) {
+                return;
+            }
+            registeredDatabaseTypes.clear();
+
+            registeredDatabaseTypes.add(new org.flywaydb.core.internal.database.cockroachdb.CockroachDBDatabaseType());
+            registeredDatabaseTypes.add(new org.flywaydb.core.internal.database.redshift.RedshiftDatabaseType());
+            registeredDatabaseTypes.add(new org.flywaydb.core.internal.database.mysql.mariadb.MariaDBDatabaseType());
+            registeredDatabaseTypes.add(new org.flywaydb.core.internal.database.db2.DB2DatabaseType());
+            registeredDatabaseTypes.add(new org.flywaydb.core.internal.database.derby.DerbyDatabaseType());
+            registeredDatabaseTypes.add(new org.flywaydb.core.internal.database.firebird.FirebirdDatabaseType());
+            registeredDatabaseTypes.add(new org.flywaydb.core.internal.database.h2.H2DatabaseType());
+            registeredDatabaseTypes.add(new org.flywaydb.core.internal.database.hsqldb.HSQLDBDatabaseType());
+            registeredDatabaseTypes.add(new org.flywaydb.core.internal.database.informix.InformixDatabaseType());
+            registeredDatabaseTypes.add(new org.flywaydb.core.internal.database.mysql.MySQLDatabaseType());
+            try {
+                // Only add the OracleDatabaseType if the driver exists
+                Class.forName("oracle.jdbc.OracleConnection");
```

Comment:
I don't like this very much but I couldn't find a safer way

Reply:
I think it's fine but let's see what @gsmet thinks

Reply:
I've also created https://github.com/flyway/flyway/pull/3163 in-lining the offending class